compiler/passes/convert-typed-uast.cpp

```diff
@@ -0,0 +1,2041 @@
+//   GOTO(epilogue)
+//   epilogue:
+// with relying on the pass-through
+void TConverter::simplifyEpilogue(FnSymbol* fn) {
+  if (fn->hasFlag(FLAG_NO_FN_BODY)) return;
+
+  if (Expr* labelDef = epilogueLabel->defPoint) {
+    if (GotoStmt* g = toGotoStmt(labelDef->prev)) {
+      if (g->gotoTag == GOTO_RETURN) {
+        // remove the GOTO and rely on fall-through
+        g->remove();
+      }
+    }
+  }
+
+  // if the epilogue label is now unused, remove it
+  if (epilogueLabel->firstSymExpr() == nullptr) {
+    epilogueLabel->defPoint->remove();
+  }
+}
+
+static ::Qualifier convertQualifier(types::QualifiedType::Kind kind) {
+  ::Qualifier q = QUAL_UNKNOWN;
+  if      (kind == types::QualifiedType::VAR)       q = QUAL_VAL;
+  else if (kind == types::QualifiedType::CONST_VAR) q = QUAL_CONST_VAL;
+  else if (kind == types::QualifiedType::CONST_REF) q = QUAL_CONST_REF;
+  else if (kind == types::QualifiedType::REF)       q = QUAL_REF;
+  else if (kind == types::QualifiedType::IN)        q = QUAL_VAL;
+  else if (kind == types::QualifiedType::CONST_IN)  q = QUAL_CONST_VAL;
+  else if (kind == types::QualifiedType::OUT)       q = QUAL_VAL;
+  else if (kind == types::QualifiedType::INOUT)     q = QUAL_VAL;
+  else if (kind == types::QualifiedType::PARAM)     q = QUAL_PARAM;
+
+  return q;
+}
+
+void TConverter::setVariableType(const uast::VarLikeDecl* v,
+                                 Symbol* sym,
+                                 RV& rv) {
+  // Get the type of the variable itself
+  if (const resolution::ResolvedExpression* rr = rv.byAstOrNull(v)) {
+    types::QualifiedType qt = rr->type();
+    if (!qt.isUnknown()) {
+      // Set a type for the variable
+      sym->type = convertType(qt.type());
+
+      // Set the Qualifier
+      auto q = convertQualifier(qt.kind());
+      if (q != QUAL_UNKNOWN)
+        sym->qual = q;
+
+      // Set the param value for the variable in paramMap, if applicable
+      /*if (sym->hasFlag(FLAG_MAYBE_PARAM) || sym->hasFlag(FLAG_PARAM)) {
+        if (qt.hasParamPtr()) {
+          Symbol* val = convertParam(qt);
+          paramMap.put(sym, val);
+        }
+      }*/
+    }
+  }
+}
+
+bool TConverter::enter(const Module* node, RV& rv) {
+  printf("enter module %s %s\n", node->id().str().c_str(), asttags::tagToString(node->tag()));
+
+  if (modulesToConvert.count(node->id()) == 0) {
+    // this module should not be converted.
+    printf("Not in modules to convert\n");
+    return false;
+  }
+
+  if (node == symbol) {
+    // we are visiting it to convert the module initialization function
+    // so proceed with the traversal
+    printf("Proceeding with module converting\n");
+    return true;
+  } else {
+    printf("Converting a submodule\n");
+
+    // it is a submodule.
+    // Set up to convert the submodule's module init function
+    // with a separate call to traverse.
+    ModuleSymbol* modSym = findOrSetupModule(node->id());
+    convertModuleInit(node, modSym);
```

Comment:
I think this is broken. My reasoning: above, you rely on `symbol`, which is a global variable. `convertModuleInit` performs "cleanup" after it's done, resetting global variables, including resetting `symbol` to `nullptr`. As such, I think it clobbers important state. That makes me think you need to preserve that kind of state and restore it (instead of clobbering).

Reply:
I think you are right that there is an issue here in order to handle submodules correctly. However I know that doesn't work (and neither do many other patterns).

In an earlier version, I was saving this state, but I convinced myself that it's better if the traversal just visits one function / module at a time. That works OK for functions and modules with the code I have here, but not for submodules. To handle those in this way, we can accumulate the submodules to visit in a list and then process them after visiting the enclosing module. I can look at doing that here.